secret_store/src/key_server_cluster/cluster_connections_net.rs

```diff
@@ -0,0 +1,543 @@
+
+			trace!(target: "secretstore_net", "{}: removing connection to {} at {}",
+				self.self_key_pair.public(), node_id, entry.get().node_address());
+			entry.remove_entry();
+		
+			true
+		} else {
+			false
+		}
+	}
+}
+
+/// Listen incoming connections.
+fn net_listen(
+	listen_address: &SocketAddr,
+	data: Arc<NetConnectionsData>,
+) -> Result<(), Error> {
+	execute(&data.executor, net_listen_future(listen_address, data.clone())?);
+	Ok(())
+}
+
+/// Listen incoming connections future.
+fn net_listen_future(
+	listen_address: &SocketAddr,
+	data: Arc<NetConnectionsData>,
+) -> Result<BoxedEmptyFuture, Error> {
+	Ok(Box::new(TcpListener::bind(listen_address)?
+		.incoming()
+		.and_then(move |stream| {
+			net_accept_connection(data.clone(), stream);
+			Ok(())
+		})
+		.for_each(|_| Ok(()))
+		.then(|_| future::ok(()))))
+}
+
+/// Accept incoming connection.
+fn net_accept_connection(
+	data: Arc<NetConnectionsData>,
+	stream: TcpStream,
+) {
+	execute(&data.executor, net_accept_connection_future(data.clone(), stream));
+}
+
+/// Accept incoming connection future.
+fn net_accept_connection_future(
+	data: Arc<NetConnectionsData>,
+	stream: TcpStream,
+) -> BoxedEmptyFuture {
+	Box::new(io_accept_connection(stream, data.self_key_pair.clone())
+		.then(move |result| net_process_connection_result(data, None, result))
+		.then(|_| future::ok(())))
+}
+
+/// Connect to remote node.
+fn net_connect(
+	data: Arc<NetConnectionsData>,
+	remote: SocketAddr,
+) {
+	execute(&data.executor, net_connect_future(data.clone(), remote));
+}
+
+/// Connect to remote node future.
+fn net_connect_future(
+	data: Arc<NetConnectionsData>,
+	remote: SocketAddr,
+) -> BoxedEmptyFuture {
+	let disconnected_nodes = data.container.disconnected_nodes();
+	Box::new(io_connect(&remote, data.self_key_pair.clone(), disconnected_nodes)
+		.then(move |result| net_process_connection_result(data, Some(remote), result))
+		.then(|_| future::ok(())))
+}
+
+/// Process network connection result.
+fn net_process_connection_result(
+	data: Arc<NetConnectionsData>,
+	outbound_addr: Option<SocketAddr>,
+	result: Result<DeadlineStatus<Result<IoConnection, Error>>, TimeoutError<io::Error>>,
+) -> IoFuture<Result<(), Error>> {
+	match result {
+		Ok(DeadlineStatus::Meet(Ok(connection))) => {
+			let connection = Arc::new(NetConnection::new(data.executor.clone(), outbound_addr.is_none(), connection));
+			if data.insert(connection.clone()) {
+				let maintain_action = data.trigger.lock().on_connection_established(connection.node_id());
+				maintain_connection_trigger(data.clone(), maintain_action);
+
+				return net_process_connection_messages(data, connection);
+			}
+		},
+		Ok(DeadlineStatus::Meet(Err(err))) => {
+			warn!(target: "secretstore_net", "{}: protocol error '{}' when establishing {} connection{}",
+				data.self_key_pair.public(), err, if outbound_addr.is_some() { "outbound" } else { "inbound" },
+				outbound_addr.map(|a| format!(" with {}", a)).unwrap_or_default());
```

Comment:
This could be `map_or/map_or_else` (String::new() doesn't allocate so I guess map_or is ok)